File: apps/mobile/lib/features/map_discovery/data/dtos/cabin_dto.dart
import "../../domain/entities/cabin.dart";
import "money_dto.dart";

/// JSON mapping for the `Cabin` schema in docs/api/openapi.yaml.
class CabinDto {
  const CabinDto({
    required this.id,
    required this.code,
    required this.type,
    required this.occupancyStatus,
    required this.isPaid,
    required this.price,
  });

  factory CabinDto.fromJson(Map<String, dynamic> json) {
    return CabinDto(
      id: json["id"] as String,
      code: json["code"] as String,
      type: json["type"] as String,
      occupancyStatus: json["occupancyStatus"] as String,
      isPaid: json["isPaid"] as bool,
      price: json["price"] == null
          ? null
          : MoneyDto.fromJson(json["price"] as Map<String, dynamic>),
    );
  }

  final String id;
  final String code;

  /// Wire value: `H`/`F`/`Slatoki`/`PMR` (ERD §3.5).
  final String type;
  final String occupancyStatus;
  final bool isPaid;
  final MoneyDto? price;

  Cabin toEntity() {
    return Cabin(
      id: id,
      code: code,
      type: switch (type) {
        "H" => CabinType.men,
        "F" => CabinType.women,
        "Slatoki" => CabinType.slatoki,
        "PMR" => CabinType.pmr,
        _ => CabinType.men,
      },
      occupancyStatus: CabinOccupancyStatus.values.firstWhere(
        (v) => v.name == occupancyStatus,
        orElse: () => CabinOccupancyStatus.outOfService,
      ),
      isPaid: isPaid,
      price: price?.toEntity(),
    );
  }
}

File: apps/backend/src/modules/station-network/interface/dto/cabin.dto.ts
import { ApiProperty } from '@nestjs/swagger';
import { Cabin } from '../../domain/entities/cabin.entity';
import { MoneyDto } from './money.dto';

/** Matches openapi.yaml components.schemas.Cabin exactly. */
export class CabinDto {
  @ApiProperty({ format: 'uuid' })
  id!: string;

  @ApiProperty()
  code!: string;

  @ApiProperty({ enum: ['H', 'F', 'Slatoki', 'PMR'] })
  type!: string;

  @ApiProperty({ enum: ['free', 'occupied', 'out_of_service'] })
  occupancyStatus!: string;

  @ApiProperty()
  isPaid!: boolean;

  @ApiProperty({ type: MoneyDto, nullable: true })
  price!: MoneyDto | null;

  static fromDomain(cabin: Cabin): CabinDto {
    const dto = new CabinDto();
    dto.id = cabin.id;
    dto.code = cabin.code;
    dto.type = cabin.type;
    dto.occupancyStatus = cabin.occupancyStatus;
    dto.isPaid = cabin.isPaid;
    dto.price = cabin.price ? MoneyDto.fromDomain(cabin.price) : null;
    return dto;
  }
}
